### PR TITLE
Tidy up old Content & Alert run database content

### PR DESCRIPTION
**Pull request by iant**

### CODE REVIEW

File: app/models/content_generation_run.rb [hunk 1]
```diff
@@ -17,17 +17,17 @@
...
-  has_many :management_priorities
-  has_many :analysis_pages
-  has_many :management_dashboard_tables
-  has_many :alert_subscription_events
+  has_many :dashboard_alerts, dependent: :delete_all
+  has_many :find_out_mores, dependent: :delete_all
+  has_many :management_priorities, dependent: :delete_all
+  has_many :analysis_pages, dependent: :delete_all
+  has_many :management_dashboard_tables, dependent: :delete_all
+  has_many :alert_subscription_events #, dependent: :delete_all
```
Comment (iant) [inline app/models/content_generation_run.rb:25]: AlertSubscriptionEvent does not have a `content_generation_run_id` column - should it or should this line be removed?  (it does have a `alert_type_rating_content_version_id`)

```
irb(main):038:0> AlertSubscriptionEvent
=> AlertSubscriptionEvent(id: integer, alert_id: integer, contact_id: integer, status: integer, communication_type: integer, message: text, created_at: datetime, updated_at: datetime, email_id: integer, alert_type_rating_content_version_id: integer, find_out_more_id: integer, unsubscription_uuid: string, priority: decimal, subscription_generation_run_id: integer)
irb(main):039:0> ManagementDashboardTable
=> ManagementDashboardTable(id: integer, content_generation_run_id: integer, alert_id: integer, alert_type_rating_content_version_id: integer, created_at: datetime, updated_at: datetime)
```

File: app/services/alerts/delete_content_generation_run_service.rb [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+    def initialize(older_than = DEFAULT_OLDER_THAN)
+      @older_than = older_than
+    end
+
+    def delete!
+      ActiveRecord::Base.transaction do
+        content_generation_runs = ContentGenerationRun.where("created_at <= ?", @older_than)
+        find_out_mores = FindOutMore.where(content_generation_run_id: content_generation_runs)
+        AlertTypeRatingContentVersion.where(id: find_out_mores.pluck(:alert_type_rating_content_version_id)).delete_all
+        find_out_mores.delete_all
```
Comment (iant) [inline app/services/alerts/delete_content_generation_run_service.rb:15]: just to double check do we need to delete associated `AlertTypeRating`, `Activitytype` and `InterventionType`s here too? e.g.
```
          find_out_more.content_version.alert_type_rating.activity_types.delete_all
          find_out_more.content_version.alert_type_rating.intervention_types.delete_all
          find_out_more.content_version.alert_type_rating.delete
```

**Comment by iant** [comment]
@ldodds I'm happy to split this into 2 prs if it would make reviewing easier?

File: app/models/content_generation_run.rb (hunk 1 above)
Comment (ldodds) [reply app/models/content_generation_run.rb:25]: That looks like a mistake to me. They belong to a SubscriptionGenerationRun

File: app/services/alerts/content_deletion_service.rb [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+
+    def initialize(older_than = DEFAULT_OLDER_THAN)
+      @older_than = older_than
+    end
+
+    def delete!
+      ActiveRecord::Base.transaction do
+        Alerts::DeleteBenchmarkRunService.new(older_than).delete!
+        Alerts::DeleteAlertGenerationRunService.new(older_than).delete!
+        Alerts::DeleteContentGenerationRunService.new(older_than).delete!
```
Comment (ldodds) [inline app/services/alerts/content_deletion_service.rb:14]: The content tables refer to the underlying Alert tables, so we should do the content deletion first

File: app/services/alerts/delete_content_generation_run_service.rb (hunk 2 above)
Comment (ldodds) [reply app/services/alerts/delete_content_generation_run_service.rb:15]: Yes, we should ensure that this doesn't end up deleting any AlertType, AlertTypeRating or AlertTypeRatingContentVersion.

**Comment by ldodds** [comment]
@iant happy to do this in 1 PR so just flag when you're ready for next review

**Comment by iant** [comment]
think this one might be ready for another review @ldodds thanks

**Review by ldodds — APPROVED** [review]
OK, next step is to add a Rake task to run the content deletion service. Can you do a PR for that, then we'll manually do a cleanup before setting it up to run regularly.

**Pull request merged**